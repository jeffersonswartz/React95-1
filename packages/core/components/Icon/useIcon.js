import { useState, useEffect } from 'react';
import icons from '@react95/icons';
import ico from 'icojs';

const MIME_TYPE = 'image/png';

function icoParse(file) {
  if (ico.isICO(file)) {
    return ico.parse(file);
  }

  return [];
}

const useIcon = ({ name, size, fallback, bpp = 4, variant = 1 }) => {
  const [data, setAvailableIcons] = useState({
    iconUrl: '',
    availableIcons: [{}],
  });

  const changeIconUrl = newUrl => {
    setAvailableIcons(iconData => ({ ...iconData, iconUrl: newUrl }));
  };

  useEffect(() => {
    async function fetchIcon() {
      const response = await fetch(icons[name]);
      const iconBuffer = await response.arrayBuffer();

      const allIcons = await icoParse(iconBuffer);

      const iconsToRender = allIcons.map(i => ({
        size: i.width,
        url: URL.createObjectURL(new Blob([i.buffer], { type: MIME_TYPE })),
        bpp: i.bpp,
      }));

      let url;
      const match = iconsToRender.filter(i => i.size === size && i.bpp === bpp);

      if (match.length > 0) {
        const finded = match[variant - 1];
        url = finded.url;
      } else if (fallback) {
        url = iconsToRender[0].url;
      }

      setAvailableIcons({ iconUrl: url, availableIcons: iconsToRender });
    }

    fetchIcon();
  }, []);

  return {
    changeIconUrl,
    iconUrl: data.iconUrl,
    availableIcons: data.availableIcons,
  };
};

export default useIcon;
