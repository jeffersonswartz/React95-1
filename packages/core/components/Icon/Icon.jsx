import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import useIcon from './useIcon';

const I = styled.i.attrs(({ url }) => ({
  style: {
    backgroundImage: url ? `url('${url}')` : 'none',
  },
}))`
  display: block;

  background-repeat: no-repeat;
  background-position: center;
  background-size: contain;

  ${({ width, height }) => `
    width: ${width}px;
    height: ${height}px;
  `}
`;

const Icon = ({
  name,
  width,
  height,
  size,
  fallback,
  bpp,
  variant,
  ...rest
}) => {
  const { iconUrl, changeIconUrl, availableIcons } = useIcon({
    name,
    size,
    fallback,
    bpp,
    variant,
  });

  useEffect(() => {
    const icon = availableIcons.find(i => i.size === size);

    changeIconUrl(icon ? icon.url : undefined);
  }, [size]);

  return (
    <I
      name={name}
      width={width || size}
      height={height || size}
      {...rest}
      url={iconUrl}
    />
  );
};

Icon.propTypes = {
  name: PropTypes.string,
  width: PropTypes.number,
  height: PropTypes.number,
  fallback: PropTypes.bool,
  size: PropTypes.number,
  bpp: PropTypes.number,
  variant: PropTypes.number,
};

Icon.defaultProps = {
  name: null,
  width: undefined,
  height: undefined,
  fallback: true,
  size: 32,
  /** Image color depth as bits per pixel. */
  bpp: 4,
  variant: 1,
};

export default Icon;
