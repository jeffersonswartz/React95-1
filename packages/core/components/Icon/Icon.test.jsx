import React from 'react';
import { renderHook } from '@testing-library/react-hooks';
import { waitRender } from '../shared/test/utils';

import useIcon from './useIcon';
import Icon from './Icon';

describe('<Icon />', () => {
  it('should match snapshot', async () => {
    const { container } = await waitRender(<Icon name="bat" />);

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with different width and height', async () => {
    const { container } = await waitRender(
      <>
        <Icon name="bat" width={10} height={10} />
        <Icon name="bat" width={20} height={20} />
        <Icon name="bat" width={30} height={30} />
      </>,
    );

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with different size', async () => {
    const { container } = await waitRender(<Icon name="bat" size={16} />);

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with different size and without a fallback', async () => {
    const { container } = await waitRender(
      <Icon name="bat" size={64} fallback={false} />,
    );

    expect(container).toMatchSnapshot();
  });
});

describe('useIcon', () => {
  it('should return the icon by given a name', async () => {
    const { result, waitForNextUpdate } = renderHook(
      ({ name, size }) => useIcon({ name, size }),
      {
        initialProps: { name: 'bat', size: 16 },
      },
    );

    expect(result.current.iconUrl).toBe('');

    await waitForNextUpdate();

    expect(result.current.iconUrl).toBeTruthy();

    expect(result.current.availableIcons).toHaveLength(3);
  });

  it('should return no iconUrl if you ommit the size', async () => {
    const { result, waitForNextUpdate } = renderHook(
      ({ name, size }) => useIcon({ name, size }),
      {
        initialProps: { name: 'bat' },
      },
    );

    expect(result.current.iconUrl).toBe('');

    await waitForNextUpdate();

    expect(result.current.iconUrl).toBe(undefined);

    expect(result.current.availableIcons).toHaveLength(3);
  });

  it('should return a fallback iconUrl by given a wrong size', async () => {
    const { result, waitForNextUpdate } = renderHook(
      ({ name, size, fallback }) => useIcon({ name, size, fallback }),
      {
        initialProps: { name: 'bat', size: 48, fallback: true },
      },
    );

    expect(result.current.iconUrl).toBe('');

    await waitForNextUpdate();

    expect(result.current.iconUrl).not.toBeNull();

    expect(result.current.availableIcons).toHaveLength(3);
  });

  it('should return different icons due to variant prop', async () => {
    const { result, waitForNextUpdate } = renderHook(() =>
      useIcon({ name: 'bat', size: 32, variant: 1 }),
    );

    const { result: result2 } = renderHook(() =>
      useIcon({ name: 'bat', size: 32, variant: 2 }),
    );

    expect(result.current.iconUrl).toBe('');

    await waitForNextUpdate();

    const iconsWithIconSize = result.current.availableIcons.filter(
      i => i.size === 32,
    );

    expect(iconsWithIconSize).toHaveLength(2);

    const [first, second] = iconsWithIconSize;

    expect(first).not.toBe(second);
    expect(result.current.iconUrl).not.toBe(result2.current.iconUrl);
  });
});
