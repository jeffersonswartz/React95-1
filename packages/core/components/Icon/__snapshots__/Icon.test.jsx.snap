// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Icon /> should match snapshot 1`] = `
.c0 {
  display: block;
  background-repeat: no-repeat;
  background-position: center;
  background-size: contain;
  width: 32px;
  height: 32px;
}

<div>
  <i
    class="c0"
    height="32"
    name="bat"
    style="background-image: url(buffer-32-4-1);"
    width="32"
  />
</div>
`;

exports[`<Icon /> should match snapshot with different size 1`] = `
.c0 {
  display: block;
  background-repeat: no-repeat;
  background-position: center;
  background-size: contain;
  width: 16px;
  height: 16px;
}

<div>
  <i
    class="c0"
    height="16"
    name="bat"
    style="background-image: url(buffer-16-4-1);"
    width="16"
  />
</div>
`;

exports[`<Icon /> should match snapshot with different size and without a fallback 1`] = `
.c0 {
  display: block;
  background-repeat: no-repeat;
  background-position: center;
  background-size: contain;
  width: 64px;
  height: 64px;
}

<div>
  <i
    class="c0"
    height="64"
    name="bat"
    style="background-image: none;"
    width="64"
  />
</div>
`;

exports[`<Icon /> should match snapshot with different width and height 1`] = `
.c0 {
  display: block;
  background-repeat: no-repeat;
  background-position: center;
  background-size: contain;
  width: 10px;
  height: 10px;
}

.c1 {
  display: block;
  background-repeat: no-repeat;
  background-position: center;
  background-size: contain;
  width: 20px;
  height: 20px;
}

.c2 {
  display: block;
  background-repeat: no-repeat;
  background-position: center;
  background-size: contain;
  width: 30px;
  height: 30px;
}

<div>
  <i
    class="c0"
    height="10"
    name="bat"
    style="background-image: url(buffer-32-4-1);"
    width="10"
  />
  <i
    class="c1"
    height="20"
    name="bat"
    style="background-image: url(buffer-32-4-1);"
    width="20"
  />
  <i
    class="c2"
    height="30"
    name="bat"
    style="background-image: url(buffer-32-4-1);"
    width="30"
  />
</div>
`;
